### ISSUE TITLE
Problems with Jedi in VS Code

### ISSUE THREAD

**Issue by DP1701** [pappasam/jedi-language-server#266]
Hi, when I switch to Jedi in VS Code, built in modules are recognised correctly but other modules (for example PyTorch) installed with pip are not. 

Details can already be read here.
[https://github.com/microsoft/vscode-python/issues/21060#issue-1668603605](https://github.com/microsoft/vscode-python/issues/21060#issue-1668603605)

Does anyone have an idea what the problem could be?

**Comment by andriati-alex** [comment]
Apparently, this is not only in vscode, but I'm using neovim and algo get the same problem. Completions for builtin libs fine, but for any external package no sign

**Comment by pappasam** [comment]
> Apparently, this is not only in vscode, but I'm using neovim and algo get the same problem. Completions for builtin libs fine, but for any external package no sign

Have you activated your virtual environment before opening Vim? I'm using neovim, just tested out completions from a 3rd party package, and things work fine for me.

**Comment by andriati-alex** [comment]
> > Apparently, this is not only in vscode, but I'm using neovim and algo get the same problem. Completions for builtin libs fine, but for any external package no sign
> 
> Have you activated your virtual environment before opening Vim? I'm using neovim, just tested out completions from a 3rd party package, and things work fine for me.

Yes, in my setup I use the Mason plugin to install the language servers, and apart from that, I have the virtual environment I work on properly activated, and yet no completions for third part packages

**Comment by regob** [comment]
I have the same issue with VSCode.
It seems when initializing the language server options the environment is not passed. If I print the environment path where [the lsp is initialized](https://github.com/pappasam/jedi-language-server/blob/2ca1f8b0eb9359b7d261cce5e1db0aebfe5c45a5/jedi_language_server/server.py#L146C23-L146C23), it's None already. So I gues this should be passed somehow.

Enabling verbose mode using

```python
sys.argv += ['-v']
sys.exit(cli())
```
when starting the language server in the [vscode extension](https://github.com/microsoft/vscode-python/blob/main/pythonFiles/run-jedi-language-server.py) the messages are printed, and I don't see VSCode passing the environment anywhere

```text

2023-11-17 11:04:44.535 [info] INFO:pygls.protocol:Language server initialized InitializeParams(... LOTS OF PARAMS ..., client_info=InitializeParamsClientInfoType(name='Visual Studio Code', version='1.84.2'), locale='en', root_path='/home/rego/.vscode-server/extensions/ms-python.python-2023.20.0', root_uri='file:///home/rego/.vscode-server/extensions/ms-python.python-2023.20.0', initialization_options={'markupKindPreferred': 'markdown', 'completion': {'resolveEagerly': False, 'disableSnippets': True}, 'diagnostics': {'enable': True, 'didOpen': True, 'didSave': True, 'didChange': True}, 'hover': {'disable': {'keyword': {'all': True}}}, 'workspace': {'extraPaths': ['/home/rego/.vscode-server/extensions/ms-python.python-2023.20.0'], 'symbols': {'maxSymbols': 0}}}, trace=<TraceValues.Off: 'off'>, work_done_token=None, workspace_folders=[WorkspaceFolder(uri='file:///home/rego/.vscode-server/extensions/ms-python.python-2023.20.0', name='ms-python.python-2023.20.0')])
```

**Comment by rohit901** [comment]
Any fix for this? I seem to have similar problem and I have described it in https://github.com/coder/code-server/issues/6619

I hope this can be fixed..

**Comment by zengziru** [comment]
I have the same question!!!

**Comment by davidhalter** [comment]
You guys have to help us somehow with what's actually going wrong. It feels like the wrong environment is detected. But why is the wrong environment detected? What is special about your installations? How many Python versions have you installed? Are you using virtualenvs? What is your editor/IDE?

Writing "I have the same question!!!" is not helpful at all. If you want us to put in work always think of us as humans, too. Try to put in your absolute best into helping us solve the problem.

**Comment by zengziru** [comment]
> You guys have to help us somehow with what's actually going wrong. It feels like the wrong environment is detected. But why is the wrong environment detected? What is special about your installations? How many Python versions have you installed? Are you using virtualenvs? What is your editor/IDE?
> 
> Writing "I have the same question!!!" is not helpful at all. If you want us to put in work always think of us as humans, too. Try to put in your absolute best into helping us solve the problem.

OK，I have solved the problem this afternoon, just download "Pyright" extention from vscode market, it can solve all problems. Because of the consideration of security, I have to use "code-server" to attach the server, but "Pylance" can not be used in code-server. So I have to use jedi, while it seems that it can not detect the virtual conda environment automatically...
So just use "Pyright", it can solve all problems.  Thank you for all efforts trying to solve the problem for almost one year.

**Comment by bbrener1** [comment]
I've run into this issue and diagnosed my particular failure mode:

When I run VSCode (well cursor, but who's counting) to connect to a remote server, individual jupyter notebooks are configured to run using a particular conda environment, but jedi itself was being invoked by the default python kernel, which was bare of any packages. 

For those who are coming back to this issue, if you are on VSCode or any of its derivatives:

Open a bare terminal IN VSCODE and invoke "which python" or run python within that terminal. Verify the path of python being invoked. If the path of the python being invoked in a bare vscode terminal is not the path of the python you are invoking for evaluation, autocompletion will only work for packages in your default path.

To fix for vscode, change the python interpreter being invoked to the one you are running the code with (or just one that has all your packages of interest in the python path) 

This is user error, but it's a subtle question of invocation order, so I'm not surprised people are coming in to this thread confused. I don't know if there is an easy way to validate and warn that jedi is being invoked from a different environment than a file is being run, since jedi in itself doesn't necessarily know what the hell you're doing with a bare .py file. That being said, jedi doesn't warn you about imports it can't find (eg when jedi is being used to analyze a file, it has no reaction to import <foo> if foo is not in the path) Similarly, jedi doesn't warn or complain if a requested signature comes back blank. I don't know if there's an easy way to make such a warning user-friendly without cluttering up the UI under other circumstances, but generally it seems like a good time to warn.
